### PR TITLE
Create the job Scheduler

### PR DESCRIPTION
**Pull request by milljm**
Create a job scheduler base class and use RunParallel as a plugin.
This will pave the way for additional scheduler plugins.

Refs #9000

### CODE REVIEW

**Review by permcody — CHANGES_REQUESTED** [review]
So even this first commit isn't quite right. On one hand you are doing really well moving most of the RunParallel class up to the top level base class from a code reusability perspective. On the other hand, it feels just wrong. Why do schedulers need to do all this spin waiting, slot checking and joining? You'd only need that if you were running jobs and not trying to overload the system. Otherwise, you just launch and go!

**Comment by milljm** [comment]
@permcody well... 'cyclic' and 'skipped dependency' detection is very closely tied to join() and the other slot management stuff. Its complex and it works. I'd like to leave it alone. But if I must then I must.

I totally agree some stuff shouldn't be in there though: Insufficient Slots, Oversized Jobs, etc. Are you asking for an elegant refactoring of all these features out of the Scheduler and into RunParallel? 

In other words, I am envisioning the Scheduler Class primarily being a 'Dependency Manager' and the handling of executing binaries (subprocess).

**Comment by permcody** [comment]
What if we didn't try to solve any dependency crap at all. Just get a list of jobs and go. Yeah if they fail other jobs may just sit in the queue but doesn't that remove all the complexity? Is that solution unworkable? It makes the queueing class dead simple.
